syncing/src/test/kotlin/maru/syncing/SyncControllerThreadSafetyTest.kt

```diff
@@ -0,0 +1,289 @@
+      }
+
+      executor.shutdown()
+      assertThat(executor.awaitTermination(30, TimeUnit.SECONDS)).isTrue()
+
+      // Verify that ELSyncStatus is never SYNCED when CL status is SYNCING
+      val finalClStatus = syncController.getCLSyncStatus()
+      val finalElStatus = syncController.getElSyncStatus()
+      val isFullInSync = syncController.isNodeFullInSync()
+
+      assertThat(isFullInSync).isEqualTo(finalClStatus == CLSyncStatus.SYNCED && finalElStatus == ELSyncStatus.SYNCED)
+      assertThat(finalElStatus == ELSyncStatus.SYNCED && finalClStatus == CLSyncStatus.SYNCING).isFalse
+
+      // Verify we received status updates (exact count may vary due to concurrency)
+      assertThat(clStatusUpdates).size().isGreaterThan(iterations / 8)
+      assertThat(elStatusUpdates).size().isGreaterThan(iterations / 8)
+    } finally {
+      if (!executor.isShutdown) {
+        executor.shutdownNow()
+      }
+    }
+  }
+
+  @Test
+  fun `should maintain EL-follows-CL invariant under concurrent access`() {
+    val executor = Executors.newFixedThreadPool(3)
+    val latch = CountDownLatch(3)
+    val iterations = 500
+    val invariantViolations = AtomicInteger(0)
+
+    // Monitor for invariant violations: when CL starts syncing, EL should never be SYNCED
+    executor.submit {
+      repeat(iterations * 3) {
+        val clStatus = syncController.getCLSyncStatus()
+        val elStatus = syncController.getElSyncStatus()
+
+        // This should never happen: CL syncing while EL is synced
+        if (clStatus == CLSyncStatus.SYNCING && elStatus == ELSyncStatus.SYNCED) {
+          invariantViolations.incrementAndGet()
+        }
+        Thread.sleep(5)
+      }
+      latch.countDown()
+    }
+
+    try {
+      // Thread 1: Rapid CL status changes
+      executor.submit {
+        repeat(iterations) { i ->
+          val status = if (i % 3 == 0) CLSyncStatus.SYNCING else CLSyncStatus.SYNCED
+          syncController.updateClSyncStatus(status)
+        }
+        latch.countDown()
+      }
+
+      // Thread 2: Rapid EL status changes
+      executor.submit {
+        repeat(iterations) { i ->
+          val status = if (i % 3 == 0) ELSyncStatus.SYNCING else ELSyncStatus.SYNCED
+          syncController.updateElSyncStatus(status)
+        }
+        latch.countDown()
+      }
+
+      assertThat(latch.await(30, TimeUnit.SECONDS)).isTrue()
+
+      // The invariant should never be violated
+      assertThat(invariantViolations.get()).isEqualTo(0)
+    } finally {
+      executor.shutdownNow()
+    }
+  }
+
+  @Test
+  fun `should handle concurrent handler invocation`() {
+    val executor = Executors.newFixedThreadPool(2)
+    val iterations = 100
+    val handlerCallCounts = AtomicInteger(0)
+    val latch = CountDownLatch(2)
+
+    // Set handlers once during initialization
+    syncController.onClSyncStatusUpdate { handlerCallCounts.incrementAndGet() }
+    syncController.onElSyncStatusUpdate { handlerCallCounts.incrementAndGet() }
```

Comment:
Can we do these as well and I guess we can at least test `beaconSyncCompletions` is equal to `iterations / 2`:
```
syncController.onBeaconSyncComplete { beaconSyncCompletions.incrementAndGet() }
syncController.onFullSyncComplete { fullSyncCompletions.incrementAndGet() }
```

Reply:
I'll add the assertion for the `beaconSyncCompletions`, but there's not much I can assert for `fullSyncCompletions ` other than there should be some